### PR TITLE
Add gradient accumulation to SentenceTransformer

### PR DESCRIPTION
**Pull request by vsuarezpaniagua**
This PR adds gradient accumulation to SentenceTransformer training, which might be very useful when using larger models or training with larger batch sizes.

*Based on #1092

### CODE REVIEW

**Comment by nreimers** [comment]
Thanks for the PR. Sadly gradient acc doesn't provide any benefits for training embedding models

**Comment by vsuarezpaniagua** [comment]
Thanks for the quick reply. However, performance in training SBERT is clearly increasing when I use larger batches for some Contrastive Learning tasks.

**Comment by nreimers** [comment]
Yes, there it works as you have more negatives in a batch. But with grad acc you miss these additional negatives. There I don't expect any impact at all.

**Comment by vsuarezpaniagua** [comment]
grad acc integrates more examples in the batch, so it actually adds more negative examples. As far as I know, having a batch size of N samples performs the same as having batches of N/M samples with M grad acc steps.

**Comment by nreimers** [comment]
Sadly not. The loss is computed with N/M negatives and then just averaged.
Eg a batch size of 512 works great, while batch size of 1 with 512 grad acc steps doesn't work at all, as you have zero negatives in each batch.

You need more clever ways to scale to larger batches for contrastive training

**Comment by vsuarezpaniagua** [comment]
In the batch size of 1 with 512, the loss is calculated for each sample and then average.

**Comment by vsuarezpaniagua** [comment]
In both cases, the final loss is calculated for each individual sample and then averaged.

**Comment by vsuarezpaniagua** [comment]
Hi @nreimers ,
I was thinking about what you said and I think there was a bit of misunderstanding. If the gradient accumulation is applied to loss functions that are pairwise computed, like ContrastiveLoss, CosineSimilarityLoss, or TripletLoss among others, it will work great because the loss calculation will be the same having M steps since it averages the individual examples. So, having the gradient accumulation implemented it would be an additional feature to the SBERT repository for training larger models or low memory GPUs. A note could be added in the comments saying that it will only work for pairwise computed losses (batch independent).

**Comment by shhshn** [comment]
Hi. Thanks for the PR!
CrossEncoder PR #1092 actually [worked for me](https://github.com/CyberAgentAILab/japanese-nli-model).
Obtained better results compared with no accumulation when you have very limited GPU memory, say 12 GB.

I am not sure whether it will work for non-regression objectives as discussed above, but can we give it a try?

File: sentence_transformers/SentenceTransformer.py [hunk 1]
```diff
@@ -710,24 +712,29 @@ def fit(self,
...
-                        optimizer.step()
+                        if (i + 1) % accumulation_steps == 0:
+                            torch.nn.utils.clip_grad_norm_(loss_model.parameters(), max_grad_norm)
+                            optimizer.step()
+                            optimizer.zero_grad()
 
-                    optimizer.zero_grad()
 
-                    if not skip_scheduler:
+                    if not skip_scheduler and (i + 1) % accumulation_steps == 0::
```
Comment (FajnyKarton) [inline sentence_transformers/SentenceTransformer.py:737]: There is a typo: '::'

**Comment by FajnyKarton** [comment]
Hi @vsuarezpaniagua 
Thanks for the PR. I am training my ST model with gradient accumulation and TripletLoss. It really helped me achieve the desired result. I definitely opt for adding it.

**Comment by thearod5** [comment]
I support this being added in. Or at least allowing an option to easily be able to toggle this on and off. Currently have to override the SentenceTransformer class and a Copy-pasted method with gradient accumulation. Would be nice to not have this be that case.

**Comment by tomaarsen** [comment]
Hello!

Since the v3.0 release, you can now use gradient accumulation with Sentence Transformers via the `gradient_accumulation_steps` [Training Argument](https://sbert.net/docs/sentence_transformer/training_overview.html#training-arguments), so I'll be closing this PR.
I do want to note that Nils' comment still holds true: gradient accumulation will not improve your performance with losses that employ in-batch negatives as there will not be more negatives to choose from. That said, it can still be used for the normal reasons.

- Tom Aarsen
